# 01. Importing Data

In [311]:
#importing data
import pandas as pd
import numpy as np
import os

In [312]:
path = r"C:\Users\andyc\Achievement 4 Project"

In [313]:
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col = False)

In [314]:
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_wrangled.csv'), index_col = False)

In [315]:
df_prods.describe()

product_id      aisle_id  department_id        prices
count  49693.000000  49693.000000   49693.000000  49693.000000
mean   24844.345139     67.770249      11.728433      9.994136
std    14343.717401     38.316774       5.850282    453.519686
min        1.000000      1.000000       1.000000      1.000000
25%    12423.000000     35.000000       7.000000      4.100000
50%    24845.000000     69.000000      13.000000      7.100000
75%    37265.000000    100.000000      17.000000     11.200000
max    49688.000000    134.000000      21.000000  99999.000000

In [316]:
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       order_day_of_week  order_hour_of_day  days_since_prior_order  
count       3.421083e+06       3.421083e+06            3.214874e+06  
mean        2.776219e+00       1.345202e+01            1.111484e+01  
std         2.046829e+00       4.226088e+00            9.206737e+00  
min         0.000000e+00       0.000000e+00            0.000000e+00  
25%         1.000000e+00       1.000000e+01            4.000000e+00  
50%         3.000000e+00       1.300000e+01            7.000000e+00  
75%         5.000000e+00       1.600000e+01            1.500000e+01  
max         6.000000e+00       2.300000e+01            3.000000e+01

In [317]:
# Create a dataframe

df_test = pd.DataFrame()

In [318]:
# Create a mixed type column

df_test['mix'] = ['a', 'b', 1, True]

In [319]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [320]:
for col in df_test.columns.tolist():
  weird = (df_test[[col]].map(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


In [321]:
df_test['mix'] = df_test['mix'].astype('str')

In [322]:
#finding missing values 
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [323]:
#creates a subset that provides all the missing values
df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [324]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [325]:
df_nan.describe()

product_id    aisle_id  department_id     prices
count     16.000000   16.000000      16.000000  16.000000
mean    6684.000000   89.937500      10.937500  13.012500
std    12836.665242   33.731229       4.639953   3.881731
min       34.000000   26.000000       1.000000   1.200000
25%      459.250000   70.750000       7.750000  12.175000
50%     2413.000000   98.500000      11.500000  13.650000
75%     3872.750000  120.000000      14.500000  14.425000
max    40440.000000  126.000000      16.000000  20.900000

### replacing missing values with the mean or median
df['column with missings'].fillna(mean value, inplace=True)
df['column with missings'].fillna(median value, inplace=True)

In [327]:
#rows and columns
df_prods.shape

(49693, 5)

In [328]:
#creating a new dataframe that shows non missing values
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [329]:
df_prods_clean.shape

(49677, 5)

In [330]:
#drop all missing values
df_prods.dropna(inplace = True)

In [331]:
df_prods.shape

(49677, 5)

In [332]:
#drops only the NaNs
df_prods.dropna(subset = ['product_name'], inplace = True)

In [333]:
df_prods.describe()

product_id      aisle_id  department_id        prices
count  49677.000000  49677.000000   49677.000000  49677.000000
mean   24850.194235     67.763110      11.728687      9.993164
std    14340.588602     38.316396       5.850651    453.592708
min        1.000000      1.000000       1.000000      1.000000
25%    12433.000000     35.000000       7.000000      4.100000
50%    24851.000000     69.000000      13.000000      7.100000
75%    37267.000000    100.000000      17.000000     11.100000
max    49688.000000    134.000000      21.000000  99999.000000

In [334]:
#looks for full duplicates in your dataframe
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [335]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [336]:
#checking the rows and colunms before deleting duplicates
df_prods_clean.shape

(49677, 5)

In [337]:
#dropping duplicates
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [338]:
df_prods_clean_no_dups

product_id                                       product_name  \
0               1                         Chocolate Sandwich Cookies   
1               2                                   All-Seasons Salt   
2               3               Robust Golden Unsweetened Oolong Tea   
3               4  Smart Ones Classic Favorites Mini Rigatoni Wit...   
4               5                          Green Chile Anytime Sauce   
...           ...                                                ...   
49688       49684          Vodka, Triple Distilled, Twist of Vanilla   
49689       49685                 En Croute Roast Hazelnut Cranberry   
49690       49686                                   Artisan Baguette   
49691       49687         Smartblend Healthy Metabolism Dry Cat Food   
49692       49688                             Fresh Foaming Cleanser   

       aisle_id  department_id  prices  
0            61             19     5.8  
1           104             13     9.3  
2            94              7     4.5  
3            38              1    10.5  
4             5             13     4.3  
...         ...            ...     ...  
49688       124              5     5.3  
49689        42              1     3.1  
49690       112              3     7.8  
49691        41              8     4.7  
49692        73             11    13.5  

[49672 rows x 5 columns]

In [339]:
df_prods_clean_no_dups.describe()

product_id      aisle_id  department_id        prices
count  49672.000000  49672.000000   49672.000000  49672.000000
mean   24850.349775     67.762442      11.728942      9.993282
std    14340.705287     38.315784       5.850779    453.615536
min        1.000000      1.000000       1.000000      1.000000
25%    12432.750000     35.000000       7.000000      4.100000
50%    24850.500000     69.000000      13.000000      7.100000
75%    37268.250000    100.000000      17.000000     11.100000
max    49688.000000    134.000000      21.000000  99999.000000

In [340]:
#finds values that are over 1000 in the prices column
df_max_prods = df_prods_clean_no_dups.loc[df_prods_clean_no_dups['prices'] > 1000]

In [341]:
df_max_prods

product_id                      product_name  aisle_id  department_id  \
21554       21553  Lowfat 2% Milkfat Cottage Cheese       108             16   
33666       33664             2 % Reduced Fat  Milk        84             16   

        prices  
21554  14900.0  
33666  99999.0

In [342]:
#replaces the specific value of 99999 with 4
df_prods_clean_no_dups = df_prods_clean_no_dups.replace(99999, 4)

In [343]:
df_prods_clean_no_dups.describe()

product_id      aisle_id  department_id        prices
count  49672.000000  49672.000000   49672.000000  49672.000000
mean   24850.349775     67.762442      11.728942      7.980176
std    14340.705287     38.315784       5.850779     66.951833
min        1.000000      1.000000       1.000000      1.000000
25%    12432.750000     35.000000       7.000000      4.100000
50%    24850.500000     69.000000      13.000000      7.100000
75%    37268.250000    100.000000      17.000000     11.100000
max    49688.000000    134.000000      21.000000  14900.000000

In [344]:
#replaces the specific value of 14900 with 3
df_prods_clean_no_dups = df_prods_clean_no_dups.replace(14900, 3)

In [345]:
df_prods_clean_no_dups.describe()

product_id      aisle_id  department_id        prices
count  49672.000000  49672.000000   49672.000000  49672.000000
mean   24850.049867     67.762442      11.728942      7.680269
std    14341.069149     38.315784       5.850779      4.199381
min        1.000000      1.000000       1.000000      1.000000
25%    12431.750000     35.000000       7.000000      4.100000
50%    24850.500000     69.000000      13.000000      7.100000
75%    37268.250000    100.000000      17.000000     11.100000
max    49688.000000    134.000000      21.000000     25.000000

I changed the prices so that it is more appropriate for a grocery store. I did research on the products before making any changes. A maximum of 25$ for the price is more reasonable than 99999 or 14900.

# 02. Exporting dataframe

In [348]:
#exporting df_prods_clean_no_dups dataframe to the preprared data folder
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data','Prepared Data', 'products_checked.csv'))

# 03. Task 4.5

### 2.)Run the df.describe() function on your df_ords dataframe.

In [351]:
#descriptive statistics
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       order_day_of_week  order_hour_of_day  days_since_prior_order  
count       3.421083e+06       3.421083e+06            3.214874e+06  
mean        2.776219e+00       1.345202e+01            1.111484e+01  
std         2.046829e+00       4.226088e+00            9.206737e+00  
min         0.000000e+00       0.000000e+00            0.000000e+00  
25%         1.000000e+00       1.000000e+01            4.000000e+00  
50%         3.000000e+00       1.300000e+01            7.000000e+00  
75%         5.000000e+00       1.600000e+01            1.500000e+01  
max         6.000000e+00       2.300000e+01            3.000000e+01

There is an unnamed column that represents the same values as order_id. I don't see anything else out of place in the data.

### 3.)Check for mixed-type data in your df_ords dataframe

In [354]:
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].map(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)

In [355]:
df_ords.dtypes

Unnamed: 0                  int64
order_id                    int64
user_id                     int64
order_number                int64
order_day_of_week           int64
order_hour_of_day           int64
days_since_prior_order    float64
dtype: object

### 4.) If you find mixed-type data, fix it. The column in question should contain observations of a single data type.

There are no mixed-type data in the df_ords dataframe.

### 5.) Run a check for missing values in your df_ords dataframe.

In [359]:
#checking for missing values in df_ords
df_ords.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
order_number                   0
order_day_of_week              0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

There is 206,209 missing values in the days_since_prior_order column. The missing values are present because there is no order before the first order of each user_id.

### 6.) Address the missing values using an appropriate method.

In [362]:
#creates a subset that provides all the missing values
df_missing_values = df_ords[df_ords['days_since_prior_order'].isnull() == True]

In [363]:
df_missing_values

Unnamed: 0  order_id  user_id  order_number  order_day_of_week  \
0                 0   2539329        1             1                  2   
11               11   2168274        2             1                  2   
26               26   1374495        3             1                  1   
39               39   3343014        4             1                  6   
45               45   2717275        5             1                  3   
...             ...       ...      ...           ...                ...   
3420930     3420930    969311   206205             1                  4   
3420934     3420934   3189322   206206             1                  3   
3421002     3421002   2166133   206207             1                  6   
3421019     3421019   2227043   206208             1                  1   
3421069     3421069   3154581   206209             1                  3   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
11                      11                     NaN  
26                      14                     NaN  
39                      11                     NaN  
45                      12                     NaN  
...                    ...                     ...  
3420930                 12                     NaN  
3420934                 18                     NaN  
3421002                 19                     NaN  
3421019                 15                     NaN  
3421069                 11                     NaN  

[206209 rows x 7 columns]

In [364]:
df_missing_values.describe()

Unnamed: 0      order_id        user_id  order_number  \
count  2.062090e+05  2.062090e+05  206209.000000      206209.0   
mean   1.712637e+06  1.708462e+06  103105.000000           1.0   
std    9.874816e+05  9.881299e+05   59527.555167           0.0   
min    0.000000e+00  2.000000e+01       1.000000           1.0   
25%    8.579730e+05  8.507300e+05   51553.000000           1.0   
50%    1.717068e+06  1.706246e+06  103105.000000           1.0   
75%    2.570518e+06  2.564292e+06  154657.000000           1.0   
max    3.421069e+06  3.421081e+06  206209.000000           1.0   

       order_day_of_week  order_hour_of_day  days_since_prior_order  
count      206209.000000      206209.000000                     0.0  
mean            2.754118          13.626597                     NaN  
std             2.076205           4.223769                     NaN  
min             0.000000           0.000000                     NaN  
25%             1.000000          11.000000                     NaN  
50%             3.000000          14.000000                     NaN  
75%             5.000000          17.000000                     NaN  
max             6.000000          23.000000                     NaN

In [365]:
df_missing_values.shape

(206209, 7)

In [366]:
df_ords.shape

(3421083, 7)

In [367]:
#creating a new dataframe that shows non missing values
df_ords_clean = df_ords[df_ords['days_since_prior_order'].isnull() == False]

In [368]:
#checks for the rows and columns for df_ords that shows non missing values
df_ords_clean.shape

(3214874, 7)

In [369]:
df_ords.shape

(3421083, 7)

I ended up not dropping the NULL values because we still want to maintain each customer's first order. The 'days_since_prior_order' column is ok to be null becaause there is not order beyond the first order.

In [371]:
df_a_customer = df_missing_values.loc[df_missing_values['user_id'] == 1]

In [372]:
df_a_customer

Unnamed: 0  order_id  user_id  order_number  order_day_of_week  \
0           0   2539329        1             1                  2   

   order_hour_of_day  days_since_prior_order  
0                  8                     NaN

After running a code to check one user_id, it makes sense that the first order for each user_id is NaN or NULL because there is no order before the first order. I can conclude that this data is still useable for other measures.

### 7.) Run a check for duplicate values in your df_ords data.

In [375]:
#looks for full duplicates in the df_ords_clean dataframe
df_dupes = df_ords_clean[df_ords_clean.duplicated()]

In [376]:
df_dupes

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, order_number, order_day_of_week, order_hour_of_day, days_since_prior_order]
Index: []

There are no duplicates found in the df_ords_clean dataframe.

### 8.) Address the duplicates using an appropriate method.

As stated above, there are no duplicates to address in the df_ords_clean dataframe.

### 9.) Export your final, cleaned df_prods and df_ords data as “.csv” files in your “Prepared Data” folder and give them appropriate, succinct names.

In [381]:
#exporting df_ords_clean dataframe to the preprared data folder
df_ords.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_checked.csv'))

We are keeping the original dataframe since we did not change anything and only checked the orders.